Revise PR template. (#10774)
rapidsai/cudf: .github/PULL_REQUEST_TEMPLATE.md

## Description

> <!-- Describe the changes. Link to any issues closed by this pull request. -->

Reviewer: ```suggestion
<!-- Standalone description of changes (will be included in CHANGELOG). Reference any issues closed by this PR with "closes #1234" or "fixes #5678" -->
```

Reviewer 2: Let's be consistently imperative here:

```
<!-- Provide a standalone description of changes in this PR (will be included in CHANGELOG). -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
```

Is the description for each PR actually included in CHANGELOG? I don't see them: https://github.com/rapidsai/cudf/blob/branch-22.06/CHANGELOG.md

Author: You're correct. The _title_ is included in the changelog. I do think that's helpful to note.

Author: Updated in 2cf27046840523ed61471bb981fe23c8d96551e5. Feel free to open a new thread for further suggestions. This one is outdated so I'll resolve it.

   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by

>    merging the target branch into the branch used for making the pull request.

Reviewer 3: This part of the advice arguably could be ported into `CONTRIBUTING.md`. This seems like useful workflow information for contributors (especially new ones). IIUC, cudf uses squash-merges for PRs, so fixing up commits in response to review is not necessary for a clean commit history (and, as noted here, can hinder in the case of multiple rounds of reivew).

Reviewer 3: I think that if it's outdated then there's no real reason to include it. I agree that github seems to be better than a few years ago at handling this kind of code movement (I don't recall having comments resolved by rebase and github losing track for a while).

Author: As I think about this more, this suggests an opportunity to rework some of the CONTRIBUTING guide. The differentiation between "[Your First Issue](https://github.com/rapidsai/cudf/blob/ad126065109aaa72b6eb324ba5abd555b70bb4ae/CONTRIBUTING.md#your-first-issue)" and "[Seasoned Developers](https://github.com/rapidsai/cudf/blob/ad126065109aaa72b6eb324ba5abd555b70bb4ae/CONTRIBUTING.md#seasoned-developers)" is irrelevant and a little presumptive. It'd be clearer to write something like "Development workflow" and "Project planning and organization", and incorporate tips about draft PRs, etc. To keep this focused, I will create a separate PR for the CONTRIBUTING guide before resolving this thread.

Reviewer 4: https://stackoverflow.com/questions/50132325/github-advises-force-pushing-can-corrupt-your-pull-request-why

Comments are not lost when you force push, but they can be disassociated with the code when the commit they refer to is lost. You can no longer see the code they are referring to when you click on the file link at the top of the comment.  

I believe it would also be possible to lose changes from other contributors or changes made via the github UI (e.g. via code review suggestions). 

This is why I recommended we have a rule against force pushing once a PR is under review.  I believe that rule/suggestion should continue to be documented.

As the above linked SO answer suggests, 
> When you really need to do a force push, you can always create a new pull request, which is much safer and keeps away annoying problems

Reviewer 4: I wish these templates and docs were RAPIDS-wide rather than specific to cuDF.

> ## Checklist
> - [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
> - [ ] New or existing tests cover these changes.
> - [ ] The documentation is up to date with these changes.

Reviewer: Two additional things came to mind. Feel free to push back on these ideas.

```suggestion
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] New or existing benchmarks cover these changes.
- [ ] The documentation is up to date with these changes.
- This change is:
   - [ ] Breaking
   - [ ] Non-breaking
```

Author: Benchmarks I can support -- with the caveat that C++ has the best benchmarking at present, and the Python is somewhat outdated.

Breaking/non-breaking is already handled by labels, and the CI requires us to label one or the other. I don't think it is necessary to input that information twice.

Reviewer 2: IMO tests are going to be much more frequent additions than benchmarks (the latter is typically for new features)

Author: I would strongly discourage duplicating information in the template and labels. The overhead of every checkbox / button adds up. However, I do agree that our categories are hard to locate. This is the official reference, but I don't think it's linked anywhere in the cuDF repository: https://docs.rapids.ai/resources/label-checker/#summary

It would be helpful if the output message said "Missing category label (bug, doc, feature request, improvement)".

Though on second thought, why is "feature request" a supported pull request category? Sounds like it should only apply to issues...

Reviewer 5: Adding to this, another idea is to add "category" to description text of the label. so that if I type "category", it will list only labels with "category" text in their description, and I can choose them easily.

Author: I like this idea. `category: doc` is longer, but it's a lot easier to distinguish the special category labels.

Reviewer 5: I am not suggesting to include "category" in label itself (also a good idea). Just in description text of the label. That is enough for search to filter "category" when we type "category".

Author: Aha! The description would be a good compromise because it doesn’t require changing bot logic.

Reviewer 5: could @rapidsai/admins or @rapidsai/cudf-admin update this? 
add "category" to the description text of the following labels: bug, doc, feature request, improvement.
This allows easily identifying the category labels during PR creation, by simply typing "category" in label box.

> - [ ] The documentation is up to date with these changes.

Author: Ooh, interesting. I hadn't considered that. I'm used to the GitHub default message for squash merges, which looks like

```
Title of PR. (https://github.com/rapidsai/cudf/pull/10774)

* First commit message

* Second commit message

* Third commit message
```

Without calling out any specific commits, the existing (long!) PR template comment can be found several times in the output of `$ git log | grep "Thank you for contributing"`. At least this is shorter?

> - [ ] The documentation is up to date with these changes.

Reviewer 6: Have you started CI issue to filter the checklist from the description? I would think that work would perhaps need to be done before merging this PR.

> - [ ] The documentation is up to date with these changes.

Reviewer 5: https://github.com/rapidsai/ops-bot/issues/102
That's not really required before this PR merge. Even if the commit text contains this checklist, it's not a big issue. Although we could request the developers to remove the checklist before doing gpucibot merge.

@bdice does it make sense to add another point to delete the checklist before merging?

> - [ ] The documentation is up to date with these changes.

Author: I would vote to take no action here, and accept that the checklist will be included in squash commit messages. It’s not really a problem in my eyes.

> - [ ] The documentation is up to date with these changes.

Author: Update: https://github.com/rapidsai/ops-bot/pull/103/ was merged, so only the contents of `## Description` will be used for the PR summary. That's a nice feature, thanks @karthikeyann / @ajschmidt8!

## Description

> <!-- Provide a standalone description of changes in this PR. -->

Reviewer 2: Sorry for the nit: "standalone", as opposed to what?

Author: @jrhemstad suggested this word (feel free to voice your thoughts, Jake). I agree it may not be needed.
```suggestion
<!-- Provide a description of changes in this PR. -->
```

Reviewer 6: > Sorry for the nit: "standalone", as opposed to what?

I think as opposed to just linking to an issue for example.

Reviewer: > I think as opposed to just linking to an issue for example.

Exactly. Far too many PRs still do this.

"Self-contained" could also be used in lieu of "standalone" if that's preferred. 

The main idea is that it shouldn't rely on referencing anything else to understand the gist of what the PR does.
